### PR TITLE
Support for stoppable process

### PR DESCRIPTION
**Pull request by kubaplas**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no

Adds support for stoppable processes to Import Definitions.
Related PR: https://github.com/dpfaffenbauer/ProcessManager/issues/37

### CODE REVIEW

File: src/DataDefinitionsBundle/Importer/Importer.php [hunk 1]
```diff
@@ -173,18 +182,67 @@ public function doImport(ImportDefinitionInterface $definition, $params)
...
+        }
+    }
+
+    /**
+     * @return bool
+     */
+    public function shouldStop()
+    {
+        if (!$this->processId) {
+            $this->processId = \Pimcore::getContainer()->get(ProcessManagerImportListener::class)->getProcess()->getId();
```
Comment (dpfaffenbauer) [inline src/DataDefinitionsBundle/Importer/Importer.php]: use di directly

File: src/DataDefinitionsBundle/Importer/Importer.php [hunk 2]
```diff
@@ -173,18 +182,67 @@ public function doImport(ImportDefinitionInterface $definition, $params)
...
+
+    /**
+     * @throws \Exception
+     */
+    public function processFailedImport(ImportDefinitionInterface $definition)
+    {
+        $this->sendDocument($definition, Document::getById($definition->getFailureNotificationDocument()),
+            $objectIds, $exceptions);
+        $this->eventDispatcher->dispatch($definition, 'data_definitions.import.failure', $params);
+        $processRepository = \Pimcore::getContainer()->get('process_manager.repository.process');
```
Comment (dpfaffenbauer) [inline src/DataDefinitionsBundle/Importer/Importer.php]: use di directly

File: src/DataDefinitionsBundle/Importer/Importer.php [hunk 3]
```diff
@@ -173,18 +182,67 @@ public function doImport(ImportDefinitionInterface $definition, $params)
...
+    /**
+     * @return bool
+     */
+    public function shouldStop()
+    {
+        if (!$this->processId) {
+            $this->processId = \Pimcore::getContainer()->get(ProcessManagerImportListener::class)->getProcess()->getId();
+        }
+
+        $processRepository = \Pimcore::getContainer()->get('process_manager.repository.process');
```
Comment (dpfaffenbauer) [inline src/DataDefinitionsBundle/Importer/Importer.php]: use di directly

**Comment by kubaplas** [comment]
Is the general solution okay for you - modyfing Process through Import Listener? There is a lot of data being saved and read, but I wanted to avoid rewriting existing models or repositories just to not overwrite one flag. Other solution would be to use another model just for statuses and stoppable flag in PM, but it would get more complex.

**Comment by dpfaffenbauer** [comment]
I think that is ok. A separate entity wouldn't make much sense here... 

I am just thinking about the Database Hits this might generate. Since the process could be stopped any time, IM needs to query that a lot, right? Maybe we define a service in PM to check in there as well, in there we can do things like use a Key-Value store to check for stuff, or use the Pimcore cache or something like that. You know what I mean?

**Comment by kubaplas** [comment]
I'm not sure if keeping the data outside the database is a good idea. In distributed installations, cache or filesystem may not be shared between instances.

On the other hand from what I see, I've added 3 read queries per import row if it is stoppable. Its impact should be minimal compared to saves that ProcessManagerImportListener does already or to the work done by Pimcore around data objects.

**Comment by kubaplas** [comment]
@dpfaffenbauer can you take a look at the code again?

File: src/DataDefinitionsBundle/ProcessManager/ProcessManagerImportListener.php [hunk 4]
```diff
@@ -38,16 +41,30 @@ final class ProcessManagerImportListener
...
     {
         $this->processFactory = $processFactory;
         $this->processLogger = $processLogger;
+        $this->repository = $repository;
     }
 
+    /**
+     * @return ProcessInterface
+     */
+    public function getProcess(): ProcessInterface
```
Comment (dpfaffenbauer) [inline src/DataDefinitionsBundle/ProcessManager/ProcessManagerImportListener.php]: Process could be null here

**Comment by dpfaffenbauer** [comment]
@kubaplas You also need to do that for exports

**Comment by kubaplas** [comment]
@dpfaffenbauer stoppable exports are here, but I want to refactor some code yet.

**Comment by dpfaffenbauer** [comment]
@kubaplas sure, just let me know when you are finished and I'll review it then ok?

**Comment by kubaplas** [comment]
@dpfaffenbauer I thought about it and I think I'm okay with the code - unless you can suggest something.

File: src/DataDefinitionsBundle/Exporter/Exporter.php [hunk 5]
```diff
@@ -73,6 +78,32 @@ final class Exporter implements ExporterInterface
      */
     private $exceptions = [];
 
+    /**
+     * @param ProcessManagerExportListener $exportListener
+     */
+    public function setExportListener(ProcessManagerExportListener $exportListener): void
```
Comment (dpfaffenbauer) [inline src/DataDefinitionsBundle/Exporter/Exporter.php]: move down

File: src/DataDefinitionsBundle/Exporter/Exporter.php [hunk 6]
```diff
@@ -360,5 +398,31 @@ private function fetchGetter(ExportMapping $map): ?GetterInterface
...
+        if (!$this->processRepository) {
+            return false;
+        }
+
+        if (!$this->processId) {
+            $this->processId = $this->exportListener->getProcess()->getId();
+        }
+
+        /** @var Process $process */
+        $process = $this->processRepository->findOneBy(['id' => $this->processId]);
```
Comment (dpfaffenbauer) [inline src/DataDefinitionsBundle/Exporter/Exporter.php]: `find`

File: src/DataDefinitionsBundle/Importer/Importer.php [hunk 7]
```diff
@@ -23,6 +23,10 @@
 use Pimcore\Model\Document;
```
Comment (dpfaffenbauer) [inline src/DataDefinitionsBundle/Importer/Importer.php]: The importer should have no idea about the Process itself, it should be decoupled. Use events to communicate.

File: src/DataDefinitionsBundle/Exporter/Exporter.php (hunk 5 above)
Comment (kubaplas) [reply src/DataDefinitionsBundle/Exporter/Exporter.php]: Moved.

File: src/DataDefinitionsBundle/Exporter/Exporter.php (hunk 6 above)
Comment (kubaplas) [reply src/DataDefinitionsBundle/Exporter/Exporter.php]: Updated.

File: src/DataDefinitionsBundle/Importer/Importer.php (hunk 7 above)
Comment (kubaplas) [reply src/DataDefinitionsBundle/Importer/Importer.php]: I removed the code from Importer and Exporter.

**Comment by kubaplas** [comment]
@dpfaffenbauer can you check the code now?

**Comment by dpfaffenbauer** [comment]
At first look, everything looks fine to me, will test that this week

**Comment by dpfaffenbauer** [comment]
thanks.

**Pull request merged**